homeassistant/helpers/config_validation.py

```diff
@@ -522,10 +522,10 @@ def template_complex(value: Any) -> Any:
             return_list[idx] = template_complex(element)
         return return_list
     if isinstance(value, dict):
-        return_dict = value.copy()
-        for key, element in return_dict.items():
-            return_dict[key] = template_complex(element)
-        return return_dict
+        return {
+            template_complex(key): template_complex(element)
+            for key, element in value.items()
+        }
     if isinstance(value, str):
         return template(value)
```

Comment:
> We would need a method that after converting something to a template, checks if there are any dynamic nodes, and if not, doesn't wrap it in a template object. Otherwise this can get very expensive.

@balloob, should that check happen here? And how does one check for dynamic nodes?

Reply:
Something like:
```
    if isinstance(value, str):
        return template(value) if "{{" in value else value
```

Reply:
What about if/else statements?

Reply:
You need to ask jinja if it contains any dynamic nodes.

Reply:
I've searched the `core` source for such a check (and Googled) but can't find how to check whether something needs to be rendered or not. Do you perhaps know how?

Reply:
https://github.com/home-assistant/core/blob/dev/homeassistant/helpers/template.py#L265

Reply:
render_to_info will gather that info.

Reply:
Do I need to use `render_to_info` or is the current implementation OK?